File: school-ms/backend/school-app/src/main/java/com/school/staff/service/ConsolidatedStaffService.java
package com.school.staff.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.school.staff.model.ConsolidatedStaff;
import com.school.staff.model.EmploymentStatus;

/**
 * Consolidated Staff service that combines functionality from both previous implementations.
 * This combines features from:
 * - com.school.hrm.service.StaffService
 * - Any other staff services in the system
 */
public interface ConsolidatedStaffService {

    /**
     * Get all staff members
     * 
     * @return List of all staff
     */
    List<ConsolidatedStaff> getAllStaff();

    /**
     * Get all staff with pagination
     * 
     * @param pageable Pagination information
     * @return Page of staff members
     */
    Page<ConsolidatedStaff> getAllStaffPaginated(Pageable pageable);

    /**
     * Get staff member by ID
     * 
     * @param id Staff ID
     * @return Staff if found, throws exception if not found
     */
    ConsolidatedStaff getStaffById(Long id);

    /**
     * Get staff member by staff ID
     * 
     * @param staffId Staff ID string
     * @return Staff if found, throws exception if not found
     */
    ConsolidatedStaff getStaffByStaffId(String staffId);

    /**
     * Get staff member by email
     * 
     * @param email Email address
     * @return Staff if found, throws exception if not found
     */
    ConsolidatedStaff getStaffByEmail(String email);

    /**
     * Get staff by role
     * 
     * @param roleId Role ID
     * @return List of staff with the given role
     */
    List<ConsolidatedStaff> getStaffByRoleId(Long roleId);

    /**
     * Get active staff
     * 
     * @return List of active staff
     */
    List<ConsolidatedStaff> getActiveStaff();
    
    /**
     * Get staff by role name
     * 
     * @param roleName Role name
     * @return List of staff with the given role
     */
    List<ConsolidatedStaff> getStaffByRoleName(String roleName);

    /**
     * Create a new staff member
     * 
     * @param staff Staff to create
     * @return Created staff with ID
     */
    ConsolidatedStaff createStaff(ConsolidatedStaff staff);

    /**
     * Update an existing staff member
     * 
     * @param id Staff ID to update
     * @param staff Updated staff information
     * @return Updated staff
     */
    ConsolidatedStaff updateStaff(Long id, ConsolidatedStaff staff);

    /**
     * Delete a staff member
     * 
     * @param id Staff ID to delete
     */
    void deleteStaff(Long id);

    /**
     * Deactivate a staff member
     * 
     * @param id Staff ID
     * @return Updated staff
     */
    ConsolidatedStaff deactivateStaff(Long id);

    /**
     * Activate a staff member
     * 
     * @param id Staff ID
     * @return Updated staff
     */
    ConsolidatedStaff activateStaff(Long id);

    /**
     * Update employment status of a staff member
     * 
     * @param id Staff ID
     * @param status New employment status
     * @return Updated staff
     */
    ConsolidatedStaff updateStaffStatus(Long id, EmploymentStatus status);
    
    /**
     * Search staff by name
     * 
     * @param query Search query
     * @return List of staff matching the query
     */
    List<ConsolidatedStaff> searchStaffByName(String query);
    
    /**
     * Get all teachers (staff with role 'TEACHER')
     * 
     * @return List of teachers
     */
    List<ConsolidatedStaff> getAllTeachers();
    
    /**
     * Bulk create staff members
     * 
     * @param staffList List of staff to create
     * @return List of created staff
     */
    List<ConsolidatedStaff> bulkCreateStaff(List<ConsolidatedStaff> staffList);
}

File: school-ms/frontend/src/services/consolidatedStaffService.ts
import { api } from './api';
import { EmploymentStatus, StaffMember } from './staffService'; // Reuse existing types

/**
 * Consolidated Staff Service that works with the new consolidated backend endpoints.
 * This service is meant to eventually replace the existing staff service.
 */
export const consolidatedStaffService = {
  // Get all staff members
  getAll: () => api.get<StaffMember[]>('staff'),
  
  // Get paginated staff
  getPaginated: (page = 0, size = 10, sortBy = 'lastName', sortDir = 'asc') => 
    api.get<{content: StaffMember[], totalElements: number, totalPages: number}>('staff/paginated', {
      params: { page, size, sortBy, sortDir }
    }),
  
  // Get a specific staff member by ID
  getById: (id: number) => api.get<StaffMember>(`staff/${id}`),
  
  // Get staff by staff ID (employee ID)
  getByStaffId: (staffId: string) => api.get<StaffMember>(`staff/staff-id/${staffId}`),
  
  // Get staff by email
  getByEmail: (email: string) => api.get<StaffMember>(`staff/email/${email}`),
  
  // Get staff by role
  getByRole: (roleName: string) => api.get<StaffMember[]>(`staff/role/${roleName}`),
  
  // Get active staff
  getActiveStaff: () => api.get<StaffMember[]>('staff/active'),
  
  // Get all teachers
  getAllTeachers: () => api.get<StaffMember[]>('staff/teachers'),
  
  // Search staff by name
  searchByName: (query: string) => api.get<StaffMember[]>('staff/search', {
    params: { query }
  }),
  
  // Create a new staff member
  create: (staffMember: StaffMember) => {
    // Ensure proper data formatting
    const formattedStaff = {
      ...staffMember,
      // Convert date strings to ISO format if needed
      joinDate: staffMember.joinDate ? new Date(staffMember.joinDate).toISOString().split('T')[0] : undefined,
      dateOfBirth: staffMember.dateOfBirth ? new Date(staffMember.dateOfBirth).toISOString().split('T')[0] : undefined
    };
    
    return api.post<StaffMember>('staff', formattedStaff);
  },
  
  // Update an existing staff member
  update: (id: number, staffMember: StaffMember) => {
    // Ensure proper data formatting
    const formattedStaff = {
      ...staffMember,
      // Convert date strings to ISO format if needed
      joinDate: staffMember.joinDate ? new Date(staffMember.joinDate).toISOString().split('T')[0] : undefined,
      dateOfBirth: staffMember.dateOfBirth ? new Date(staffMember.dateOfBirth).toISOString().split('T')[0] : undefined
    };
    
    return api.put<StaffMember>(`staff/${id}`, formattedStaff);
  },
    // Update staff employment status
  updateEmploymentStatus: (id: number, status: EmploymentStatus) => 
    api.put<StaffMember>(`staff/${id}/status?status=${status}`),
  
  // Activate a staff member
  activateStaff: (id: number) => api.put<StaffMember>(`staff/${id}/activate`),
  
  // Deactivate a staff member
  deactivateStaff: (id: number) => api.put<StaffMember>(`staff/${id}/deactivate`),
  
  // Delete a staff member
  delete: (id: number) => api.delete(`staff/${id}`),
  
  // Bulk create staff members
  bulkCreate: (staffMembers: StaffMember[]) => {
    // Format dates in the staff data
    const formattedStaff = staffMembers.map(staff => ({
      ...staff,
      joinDate: staff.joinDate ? new Date(staff.joinDate).toISOString().split('T')[0] : undefined,
      dateOfBirth: staff.dateOfBirth ? new Date(staff.dateOfBirth).toISOString().split('T')[0] : undefined
    }));
    
    return api.post<StaffMember[]>('staff/bulk-upload', formattedStaff);
  },
  
  // Get staff statistics
  getStats: () => api.get<{
    totalStaff: number,
    activeStaff: number,
    inactiveStaff: number,
    roleDistribution: Record<string, number>,
    statusDistribution: Record<string, number>
  }>('staff/stats')
};

export default consolidatedStaffService;
